src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,487 @@ extern "C" {
 using namespace gpdxl;
 using namespace gpmd;
 
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GroupingListContainsPrimaryKey
+//
+//	@doc:
+//		Checks if list grouping_list contains all columns from conkeys of
+//		Primary Key in relation conrelid. grouping_list is treated as flat
+//		list of SortGroupClauses, other nodes are ignored.
+//		Important note: grouping_list should not have duplicate values.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
+											   List *grouping_list,
+											   List *conkeys, Oid conrelid)
+{
+	ListCell *lgc;
+	ULONG found_col_cnt = 0;
+
+	// Check if conkeys set is a subset of grouping columns set.
+	ForEach(lgc, grouping_list)
+	{
+		Node *grpcl = (Node *) lfirst(lgc);
+
+		if (NULL == grpcl || !IsA(grpcl, SortGroupClause))
+		{
+			continue;
+		}
+
+		SortGroupClause *sgc = (SortGroupClause *) grpcl;
+		Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
+
+		if (!IsA(expr, Var))
+		{
+			continue;
+		}
+
+		Var *var = (Var *) expr;
+
+		RangeTblEntry *rte =
+			(RangeTblEntry *) gpdb::ListNth(query->rtable, var->varno - 1);
+
+		if (0 == var->varlevelsup && rte->relid == conrelid &&
+			gpdb::ListMemberInt(conkeys, var->varattno))
+		{
+			found_col_cnt++;
+		}
+	}
+
+	// check that all conkeys were found
+	return (gpdb::ListLength(conkeys) == found_col_cnt);
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::AddMissingGroupClauseMutator
+//
+//	@doc:
+//		Traverses the groupClause, adds new SortGroupClause (defined by m_gc
+//		field in context) to all grouping lists where m_gc has functional
+//		dependency.
+//---------------------------------------------------------------------------
+Node *
+CQueryMutators::AddMissingGroupClauseMutator(
+	Node *node, SContextFixGroupDependentTargets *context)
+{
+	if (NULL == node)
+	{
+		return NULL;
+	}
+
+	if (IsA(node, SortGroupClause))
+	{
+		if (context->m_parent_is_grouping_clause)
+		{
+			// In case the SortGroupClause is direct element of groupsets list
+			// of GroupingClause, we treat it in a special way - consider
+			// it like a 1-element list.
+			List *ls = ListMake1(node);
+			Node *new_node = AddMissingGroupClauseMutator((Node *) ls, context);
+			gpdb::ListFree(ls);
+			return new_node;
+		}
+		return (Node *) gpdb::CopyObject(node);
```

Comment:
because it is a proper implementation of mutator according to its documentation:
```
 * expression_tree_mutator() is designed to support routines that make a
 * modified copy of an expression tree, with some nodes being added,
 * removed, or replaced by new subtrees.  The original tree is (normally)
 * not changed.  Each recursion level is responsible for returning a copy of
 * (or appropriately modified substitute for) the subtree it is handed.
 ```

Your proposal contradicts the design concept of `expression_tree_mutator`.